From 7ef62f950ca1a7492a4fb7f0abb6afa9a3cffdb6 Mon Sep 17 00:00:00 2001
From: Octavian Purdila <octavian.purdila@intel.com>
Date: Mon, 20 Feb 2012 09:42:15 +0200
Subject: bootchart: add threads support

Grab information about threads not only processes. This gives us
additional important information about the boot processes especially
for heavy processes such as system server and zygote.

Issue: ANDROID-609

Change-Id: Ia941da116de1589c46101f262db64b0bc730e643
Signed-off-by: Octavian Purdila <octavian.purdila@intel.com>
---
 init/bootchart.c |   26 +++++++++++++++++++++-----
 1 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/init/bootchart.c b/init/bootchart.c
index f72fcaa..28ba9eb 100644
--- a/init/bootchart.c
+++ b/init/bootchart.c
@@ -239,12 +239,22 @@ do_log_file(FileBuff  log, const char*  procfile)
 }
 
 static void
-do_log_procs(FileBuff  log)
+do_log_procs(FileBuff  log, int old_pid)
 {
-    DIR*  dir = opendir("/proc");
+    DIR*  dir;
     struct dirent*  entry;
 
-    do_log_uptime(log);
+    if (!old_pid) {
+      dir = opendir("/proc");
+      do_log_uptime(log);
+    } else {
+      char tmp[1024];
+      snprintf(tmp, sizeof(tmp), "/proc/%d/task", old_pid);
+      dir = opendir(tmp);
+    }
+
+    if (!dir)
+      return;
 
     while ((entry = readdir(dir)) != NULL) {
         /* only match numeric values */
@@ -257,6 +267,9 @@ do_log_procs(FileBuff  log)
             int   len;
             int   fd;
 
+            if (pid == old_pid)
+                continue;
+
             /* read command line and extract program name */
             snprintf(filename,sizeof(filename),"/proc/%d/cmdline",pid);
             proc_read(filename, cmdline, sizeof(cmdline));
@@ -285,10 +298,13 @@ do_log_procs(FileBuff  log)
                     }
                }
             }
+            if (old_pid == 0)
+              do_log_procs(log, pid);
         }
     }
+    if (!old_pid)
+      do_log_ln(log);
     closedir(dir);
-    do_log_ln(log);
 }
 
 static FileBuffRec  log_stat[1];
@@ -355,7 +371,7 @@ int  bootchart_step( void )
 {
     do_log_file(log_stat,   "/proc/stat");
     do_log_file(log_disks,  "/proc/diskstats");
-    do_log_procs(log_procs);
+    do_log_procs(log_procs, 0);
 
     /* we stop when /data/bootchart-stop contains 1 */
     {
-- 
1.7.4.1

